remote-storage-managers/hdfs/src/main/java/org/apache/kafka/rsm/hdfs/HDFSRemoteStorageManager.java

```diff
@@ -17,124 +17,244 @@
+    public static final String TXN_INDEX_FILE_NAME = "txn";
+    public static final String PRODUCER_SNAPSHOT_FILE_NAME = "snapshot";
 
     private URI fsURI = null;
     private String baseDir = null;
     private Configuration hadoopConf = null;
-    private ThreadLocal<FileSystem> fs = new ThreadLocal<>();
+    private final ThreadLocal<FileSystem> fs = new ThreadLocal<>();
     private int cacheLineSize;
     private LRUCache readCache;
 
+    /**
+     * Initialize this instance with the given configs
+     *
+     * @param configs Key-Value pairs of configuration parameters
+     */
     @Override
-    public void copyLogSegment(RemoteLogSegmentMetadata remoteLogSegmentMetadata, LogSegmentData logSegmentData) throws RemoteStorageException {
-        try {
-            String desDir = getSegmentRemoteDir(remoteLogSegmentMetadata.remoteLogSegmentId());
+    public void configure(Map<String, ?> configs) {
+        HDFSRemoteStorageManagerConfig conf = new HDFSRemoteStorageManagerConfig(configs, true);
 
-            File logFile = logSegmentData.logSegment();
-            File offsetIdxFile = logSegmentData.offsetIndex();
-            File tsIdxFile = logSegmentData.timeIndex();
+        fsURI = URI.create(conf.getString(HDFSRemoteStorageManagerConfig.HDFS_URI_PROP));
+        baseDir = conf.getString(HDFSRemoteStorageManagerConfig.HDFS_BASE_DIR_PROP);
+        cacheLineSize = conf.getInt(HDFSRemoteStorageManagerConfig.HDFS_REMOTE_READ_MB_PROP) * 1048576;
+        long cacheSize = conf.getInt(HDFSRemoteStorageManagerConfig.HDFS_REMOTE_READ_CACHE_MB_PROP) * 1048576L;
+        if (cacheSize < cacheLineSize) {
+            throw new IllegalArgumentException(HDFSRemoteStorageManagerConfig.HDFS_REMOTE_READ_MB_PROP +
+                    " is larger than " + HDFSRemoteStorageManagerConfig.HDFS_REMOTE_READ_CACHE_MB_PROP);
+        }
+        readCache = new LRUCache(cacheSize);
 
-            FileSystem fs = getFS();
-            fs.mkdirs(new Path(desDir));
+        // Loads configuration from hadoop configuration files in class path
+        hadoopConf = new Configuration();
+    }
 
-            // copy local files to remote temporary directory
-            copyFile(fs, logFile, getPath(desDir, LOG_FILE_NAME));
-            copyFile(fs, offsetIdxFile, getPath(desDir, OFFSET_INDEX_FILE_NAME));
-            copyFile(fs, tsIdxFile, getPath(desDir, TIME_INDEX_FILE_NAME));
+    @Override
+    public void copyLogSegment(RemoteLogSegmentMetadata metadata, LogSegmentData segmentData) throws RemoteStorageException {
+        try {
+            final LogSegmentDataHeader header = LogSegmentDataHeader.create(segmentData);
+            final Path dirPath = new Path(getSegmentRemoteDir(metadata.remoteLogSegmentId()));
+            final FSDataOutputStream fsOut = getFS().create(dirPath);
+
+            fsOut.write(LogSegmentDataHeader.serialize(header), 0, LogSegmentDataHeader.LENGTH);
```

Comment:
Why are we passing `LogSegmentDataHeader.LENGTH`. Should not this be same as serialized bytes length?

Reply:
It's the same as the serialized header bytes length. Updated it for better clarity.